### PR TITLE
Send readdir as a batched request, where the response comes with a list of dirs, instead of just one.

### PR DESCRIPTION
**Pull request by meowjesty**
- Issue: #2611 

- Adds a new pair of messages to the protocol: `ReadDirBatchRequest` and `ReadDirBatchResponse`;
  - `ReadDirBatchRequest` is sent from the intproxy to the agent when the intproxy receives a `ReadDirRequest` (the layer doesn't deal with the _batch_ stuff). It gets an iterator of `InternalDirEntry` from the agent (where previously this iterator was kept in the agent only);
  - `ReadDirBatchResponse` sends the iterator to the intproxy, and the length is set in the intproxy;

### CODE REVIEW

File: mirrord/agent/src/file.rs [hunk 1]
```diff
@@ -727,6 +721,17 @@ impl FileManager {
         Ok(result)
     }
 
+    #[tracing::instrument(level = Level::DEBUG, skip(self), err)]
+    pub(crate) fn read_dir_batch(&mut self, fd: u64) -> RemoteResult<ReadDirBatchResponse> {
+        let result = self
+            .get_dir_stream(fd)?
+            .map(DirEntryInternal::try_from)
+            .try_collect::<Vec<_>>()
+            .map(|dir_entries| ReadDirBatchResponse { fd, dir_entries })?;
```
Comment (Razz4780) [inline mirrord/agent/src/file.rs:730]: Do we want to read the *whole* directory? A directory on ext4 can have like 64k entries

Comment (aviramha) [reply mirrord/agent/src/file.rs:730]: +1, I'd expect some buffering

Comment (meowjesty) [reply mirrord/agent/src/file.rs]: Changed it! Now it's being done in steps. Should I add the `amount` we read as config somewhere?

Comment (Razz4780) [reply mirrord/agent/src/file.rs]: Nah, I think we can enable configuration later if at all

File: mirrord/protocol/src/file.rs [hunk 2]
```diff
@@ -385,11 +385,23 @@ pub struct ReadDirRequest {
...
+
 #[derive(Encode, Decode, Debug, PartialEq, Eq, Clone)]
 pub struct ReadDirResponse {
     pub direntry: Option<DirEntryInternal>,
 }
 
+#[derive(Encode, Decode, Debug, PartialEq, Eq, Clone)]
+pub struct ReadDirBatchResponse {
+    pub fd: u64,
+    pub dir_entries: Option<Vec<DirEntryInternal>>,
```
Comment (Razz4780) [inline mirrord/protocol/src/file.rs:402]: We could probably skip the `Option`? Empty `Vec` means no more entries anyway

File: tests/go-e2e-dir/main.go [hunk 3]
```diff
@@ -1,3 +1,4 @@
+// TODO(alex) [high]: Change the integration to readdirbatchrequest
```
Comment (Razz4780) [inline tests/go-e2e-dir/main.go:1]: todoo

File: mirrord/layer/tests/apps/issue1123/issue1123.rs [hunk 4]
```diff
@@ -1,3 +1,4 @@
+// TODO(alex) [high]: Remote resource out of sync, is this the `None`?
```
Comment (Razz4780) [inline mirrord/layer/tests/apps/issue1123/issue1123.rs:1]: todoo

File: mirrord/layer/tests/apps/env_bash_cat.sh [hunk 5]
```diff
@@ -14,4 +14,4 @@
 # cat is a SIPed binary on mac.
 cat /very_interesting_file
 # sleep so we get close request (else bash might exit before we have time to close it)
-sleep 1
\ No newline at end of file
+sleep 10
```
Comment (Razz4780) [inline mirrord/layer/tests/apps/env_bash_cat.sh:17]: Why? :eyes:

File: mirrord/layer/src/lib.rs [hunk 6]
```diff
@@ -483,7 +483,7 @@ fn sip_only_layer_start(mut config: LayerConfig, patch_binaries: Vec<String>) {
 /// - `enabled_remote_dns`: replaces [`libc::getaddrinfo`] and [`libc::freeaddrinfo`] when this is
 ///   `true`, see [`NetworkConfig`](mirrord_config::feature::network::NetworkConfig), and
 ///   [`hooks::enable_socket_hooks`](socket::hooks::enable_socket_hooks).
-#[mirrord_layer_macro::instrument(level = "trace")]
+// #[mirrord_layer_macro::instrument(level = Level::TRACE)]
```
Comment (Razz4780) [inline mirrord/layer/src/lib.rs:486]: todoo

File: mirrord/intproxy/protocol/src/lib.rs [hunk 7]
```diff
@@ -372,6 +365,13 @@ impl_request!(
     res_path = ProxyToLayerMessage::File => FileResponse::ReadDir,
 );
 
+impl_request!(
+    req = ReadDirBatchRequest,
+    res = RemoteResult<ReadDirBatchResponse>,
+    req_path = LayerToProxyMessage::File => FileRequest::ReadDirBatch,
+    res_path = ProxyToLayerMessage::File => FileResponse::ReadDirBatch,
+);
```
Comment (Razz4780) [inline mirrord/intproxy/protocol/src/lib.rs:373]: This is required only for the code to compile, right? Layer should not send batch requests

File: mirrord/intproxy/src/proxies/simple.rs [hunk 8]
```diff
@@ -55,8 +80,15 @@ impl BackgroundTask for SimpleProxy {
     type MessageOut = ProxyMessage;
 
     async fn run(mut self, message_bus: &mut MessageBus<Self>) -> Result<(), RequestQueueEmpty> {
+        let mut protocol_version = VERSION.clone();
```
Comment (Razz4780) [inline mirrord/intproxy/src/proxies/simple.rs:83]: Should be initialized to `None`. Agent version might be lower and might not allow for batched readdir

File: mirrord/intproxy/src/proxies/simple.rs [hunk 9]
```diff
@@ -27,20 +32,40 @@ pub enum SimpleProxyMessage {
...
+#[derive(Clone)]
+pub(crate) enum FileResource {
+    File,
+    Dir {
+        dirs_iter: Option<IntoIter<DirEntryInternal>>,
+    },
+}
+
+impl FileResource {
+    fn next_dir(&mut self) -> Option<DirEntryInternal> {
```
Comment (Razz4780) [inline mirrord/intproxy/src/proxies/simple.rs:53]: This should return an error when called on a regular file

File: mirrord/intproxy/src/proxies/simple.rs [hunk 10]
```diff
@@ -85,6 +117,45 @@ impl BackgroundTask for SimpleProxy {
...
+                                    ReadDirResponse {
+                                        direntry: Some(dir),
+                                    },
+                                ))),
+                                layer_id,
+                            })
+                            .await;
+                    } else {
+                        self.file_reqs.insert(message_id, layer_id);
+                        // let protocol_version = Version::new(0, 0, 0);
```
Comment (Razz4780) [inline mirrord/intproxy/src/proxies/simple.rs:143]: Delet :wastebasket:

**Review by Razz4780 — CHANGES_REQUESTED** [review]

File: mirrord/intproxy/src/proxies/simple.rs [hunk 11]
```diff
@@ -85,6 +117,45 @@ impl BackgroundTask for SimpleProxy {
...
+                        let request = if protocol_version >= Version::new(1, 9, 0) {
```
Comment (Razz4780) [inline mirrord/intproxy/src/proxies/simple.rs:144]: This minimal version should be a static defined in `mirrord_protocol` crate

**Review by Razz4780 — CHANGES_REQUESTED** [review]
One more thing - could you add a UT to `SimpleProxy` to check the new messages? Should check readdir behavior when agent protocol version (does not) match version requirement.

File: mirrord/protocol/src/codec.rs [hunk 12]
```diff
@@ -121,6 +115,7 @@ pub enum FileResponse {
     Xstat(RemoteResult<XstatResponse>),
     XstatFs(RemoteResult<XstatFsResponse>),
     ReadDir(RemoteResult<ReadDirResponse>),
+    ReadDirBatch(RemoteResult<ReadDirBatchResponse>),
```
Comment (Razz4780) [inline mirrord/protocol/src/codec.rs:118]: This **must** be the last variant in the enum

File: mirrord/protocol/src/codec.rs [hunk 13]
```diff
@@ -79,6 +72,7 @@ pub enum FileRequest {
     XstatFs(XstatFsRequest),
     FdOpenDir(FdOpenDirRequest),
     ReadDir(ReadDirRequest),
+    ReadDirBatch(ReadDirBatchRequest),
```
Comment (Razz4780) [inline mirrord/protocol/src/codec.rs:75]: This **must** be the last variant of the enum

**Review by Razz4780 — CHANGES_REQUESTED** [review]

File: mirrord/layer/tests/apps/env_bash_cat.sh (hunk 5 above)
Comment (meowjesty) [reply mirrord/layer/tests/apps/env_bash_cat.sh]: Was testing stuff.

File: mirrord/intproxy/src/proxies/simple.rs (hunk 9 above)
Comment (meowjesty) [reply mirrord/intproxy/src/proxies/simple.rs]: Changed it to return proper errors when we hit an error in the cached readdir.

File: mirrord/intproxy/protocol/src/lib.rs (hunk 7 above)
Comment (meowjesty) [reply mirrord/intproxy/protocol/src/lib.rs:373]: Should I break the pattern? Kept it like this to have it looking like all the other requests.

File: mirrord/protocol/src/file.rs [hunk 14]
```diff
@@ -5,11 +5,18 @@ use std::fs::DirEntry;
...
+};
 
 use bincode::{Decode, Encode};
 #[cfg(target_os = "linux")]
 use nix::sys::statfs::Statfs;
+use semver::VersionReq;
+
+/// Minimal mirrord-protocol version that allows [`ReadDirBatchRequest`].
+pub static READDIR_BATCH_VERSION: LazyLock<VersionReq> =
+    LazyLock::new(|| ">=1.8.2".parse().expect("Bad Identifier"));
```
Comment (Razz4780) [inline mirrord/protocol/src/file.rs:19]: Gotta bump the crate version as well

File: mirrord/intproxy/src/proxies/simple.rs [hunk 15]
```diff
@@ -27,20 +34,103 @@ pub enum SimpleProxyMessage {
...
+
+impl FileResource {
+    fn next_dir(&mut self, remote_fd: u64) -> Result<Option<DirEntryInternal>, FileError> {
+        match self {
+            FileResource::Dir { dirs_iter } => dirs_iter.next().map(Ok).transpose(),
+            FileResource::File => Err(FileError::DirOnFile(remote_fd)),
+        }
+    }
+
+    async fn handle_readdir(
```
[the remark this answers is not in the document]
Comment (meowjesty) [reply mirrord/intproxy/src/proxies/simple.rs:96]: Moved it there.

File: mirrord/protocol/src/file.rs (hunk 14 above)
Comment (meowjesty) [reply mirrord/protocol/src/file.rs]: Should this be an `x.x.+` bump, or a `x.+.0`?

Comment (Razz4780) [reply mirrord/protocol/src/file.rs]: I think `x.+.0`, it's a new feature in the protocol

File: mirrord/protocol/src/file.rs [hunk 16]
```diff
@@ -385,11 +392,31 @@ pub struct ReadDirRequest {
...
 pub struct ReadDirResponse {
     pub direntry: Option<DirEntryInternal>,
 }
 
+/// `readdir` response with the list of items (length depends on the [`ReadDirBatchRequest`]'s
+/// `amount`), and the `remote_fd` of the dir (for convenience).
+#[derive(Encode, Decode, Debug, PartialEq, Eq, Clone)]
+pub struct ReadDirBatchResponse {
+    /// Remote fd of the dir.
+    pub fd: u64,
```
Comment (Razz4780) [inline mirrord/protocol/src/file.rs:414]: What do we need this for? Agent handles file requests sequentially, so file responses don't usually contain the descriptor. Just a question though, I don't mind if it stays here

**Review by Razz4780 — APPROVED** [review]
:cat2:

**Review by gememma — APPROVED** [review]
noice 👍

File: mirrord/protocol/src/file.rs (hunk 16 above)
Comment (meowjesty) [reply mirrord/protocol/src/file.rs:414]: Was the easiest way I found to updating the iterator:

```rust
                    if let Some(FileResource::Dir { dirs_iter }) =
                        self.remote_fds.get_mut(&layer_id, &RemoteFd::Dir(fd))
                    {
                        *dirs_iter = entries_iter;
                    }

```

Comment (Razz4780) [reply mirrord/protocol/src/file.rs:414]: Ah, okok

**Pull request merged**